Feature : Item Name Scrolling

SchemingWeasels opened the pull request:
## Issues
 - Closes #85 

## Description
Long application names now scroll on the screen after a delay.

## Types of changes
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] Requested changes are in a branch
- [X] Compiled and tested requested changes on target hardware (PC, device)
- [ ] Updated the documentation, if necessary
- [ ] Updated the LICENSES file, if necessary
- [X] Reviewed the [guidelines for contributing](../CONTRIBUTING.md) to this repository


## Further comments

Commits include changes from #125 as it hasn't been merged into master, and is a required dependency for this change.

SchemingWeasels committed:
Added support for name scrolling

Long program names now scroll on the screen.

Fixed some code commenting errors.

Application names from the desktop app come padded with extra spaces. The 'end of name' is assumed to be when there are 2 spaces back to back.
This will be changed when the desktop app has been updated.

SchemingWeasels committed:
Merge branch 'enhanced-acceleration' into feature_oled-scrolling

SchemingWeasels committed:
Removed obsolete delay code

SchemingWeasels committed:
Added acceleration setting in desktop app

SchemingWeasels committed:
Revert to Rotary library & reworked acceleration customization.

SchemingWeasels committed:
Updated item name padding

SchemingWeasels commented:
There a bug with game edit mode. If you wake the screen, sometimes the scrolling just messes everything up. Can't seem to figure out why.

t3knomanzer commented:
I haven't looked at the implementation, I'll leave that to @XScorpion2, but I pulled the code and ran it. 
Man, it's so cool to see stuff moving on the screen! I did up the steps to 30 to slow it down a bit, I like it a lot better like that.

XScorpion2 reviewed (approved):
First approach looks good. There is a bit of work / rework on how the display code is handled to really make it scale as currently the full frame is being updated every loop, which is not necessary as the oled itself has a buffer that we only need to push changes to instead of the full frame.

SchemingWeasels commented:
> There a bug with game edit mode. If you wake the screen, sometimes the scrolling just messes everything up. Can't seem to figure out why.

I would hold on merging till this is figured out. Although I have some work that needs some rather urgent attention for the next few days so I apologize if I'm blocking/holding anyone or anything up.

SchemingWeasels committed:
Merge branch 'enhanced-acceleration' into feature_oled-scrolling

t3knomanzer commented:
@XScorpion2 I don't see any method in the Adafruit SSD1306 library to do partial updates of the screen.
We can certainly send commands and data so it'd be a matter of just creating our own partial update function and calling that one instead of display->display();

Ref: 
https://electronics.stackexchange.com/questions/335152/ssd1306-i2c-draw-only-changes-works-however-on-wrong-position-how-to-set-posit

https://github.com/adafruit/Adafruit_SSD1306

XScorpion2 commented:
@t3knomanzer ya, would need to add that function. 
And as it just so has it, I did just that for the qmk keyboard firmware: 
https://github.com/qmk/qmk_firmware/blob/master/drivers/oled/oled_driver.c 
I should be able to take that and modify it for use with Arduino's wire library.

t3knomanzer commented:
@XScorpion2 That's fantastic! Actually, on another note... we should also start taking all those other .ino files and making them into real C libraries.  
I'm not super experienced in C/C++ so there is a lot of room for improvement in many areas. From code organization to style and memory management.  
I'm definitely going to take a look at the qmk_firmware and use that as reference. I'm glad the oled driver is a pure C library and not C++. IMO, C++ is overkill for MCUs but that's just an opinion.

SchemingWeasels committed:
Merge branch 'master' into feature_oled-scrolling

t3knomanzer commented:
@XScorpion2 Do you have any idea when you'll be able to tackle this? It's the only task left to release 1.1.0. No pressure of course, we all have day jobs :D

XScorpion2 commented:
I've set aside some time this weekend to work on it. Not sure if the right approach is to modify the qmk oled code, or to modify the adafruit library at this point, but either way it should be about the same amount of time to knock out.

t3knomanzer commented:
Awesome!
I really like the code in QMK.   
I personally would love to go that way. Get rid of all the third party libs we’re using and and make our own C implementations but that’s just personal preference.   
We can do one at a time of course. 

By the looks of it, the oled driver in qmk is a full implementation, completely standalone, right?

SchemingWeasels commented on Desktop/Application/MaxMix/MaxMix.csproj:
```diff
@@ -85,20 +85,20 @@
-    <Reference Include="Xceed.Wpf.Toolkit, Version=3.5.0.0, Culture=neutral, PublicKeyToken=3e4669d2f30244f4, processorArchitecture=MSIL">
-      <HintPath>..\packages\Extended.Wpf.Toolkit.3.5.0\lib\net40\Xceed.Wpf.Toolkit.dll</HintPath>
+    <Reference Include="Xceed.Wpf.Toolkit, Version=4.0.0.0, Culture=neutral, PublicKeyToken=3e4669d2f30244f4, processorArchitecture=MSIL">
+      <HintPath>..\packages\Extended.Wpf.Toolkit.4.0.1\lib\net40\Xceed.Wpf.Toolkit.dll</HintPath>
```
Ugh this is still haunting me haha

SchemingWeasels committed:
Revert legacy changes

SchemingWeasels commented:
Just to confirm, are you guys getting the crashes while in game edit mode with this PR? I've noticed that it happens a lot more with "Steam Client Bootstrapper"

t3knomanzer commented:
@SchemingWeasels I'm seeing this.
![](https://i.imgur.com/RV86Mae.jpg)

SchemingWeasels commented:
Thanks. Pulling my hair out at the moment. The drawing code is identical to the larger font size (Nav, App, Game Select Screens) but this crashes while the other doesn't.

I doubt it's font size related. App edit is also font size 1 and doesn't crash.

Perhaps it's a ram issue... my code is hella sloppy and game edit mode draws tons of stuff. 

Edit: DOH~! I didn't pass by reference in DisplayGameEditScreen. That should do it...

t3knomanzer commented:
I tried reducing the max number of items to see if we were running out of memory but doesn't seem to be the case.

I'm working on something else now but I can help you debug it over the weekend maybe.

SchemingWeasels commented:
> I tried reducing the max number of items to see if we were running out of memory but doesn't seem to be the case.
> 
> I'm working on something else now but I can help you debug it over the weekend maybe.

No worries. I figured it out. You can see the progression of catching the bug in my previous edited comment 🤦.

1. What values do you use for the step interval delay and do you have a suggestion for a default value? 

2. Should this be editable in the app? I think it is unnecessary. 

3. Do you think we should raise the default sleep time because the display hardly has time to scroll before it goes to sleep with the current default sleep time.

t3knomanzer commented:
> I did up the steps to 30 to slow it down a bit, I like it a lot better like that.

I can't remember what I increased the sleep to. You choice :D

SchemingWeasels commented:
I'll push the fix tomorrow. I still have to work a bit on the scrolling in game mode because we are going to scroll 2 items. Haven't decided on the expected behavior yet.

t3knomanzer commented:
Sounds good! Thanks for the good work @SchemingWeasels

t3knomanzer commented:
Just to be clear, we are waiting on 2 things here, the fix from @SchemingWeasels and the display code from @XScorpion2.  
Or should we just release it with the fix from @SchemingWeasels and then release the optimizations later?  

I think it all depends on 2 things, how bad is the performance with the current display update code and what does @XScorpion2 have in mind.  
Obviously if it's unsusable performance wise we can't release it but that's not what I have observed in the little testing I've done.  
As far as the update code, I'd love to hear opinions on completely switching the Adafruit libraries for the one @XScorpion2 wrote.  
My vote is on switching it completely, I really like the approach of going pure C. I'd love to slowly get rid of all third party libraries we are using and switching them for C implementations.  
Then once we have switched the libraries we can start making the multiple .ino files we have into libraries too.  
Ultimately completely getting rid of the arduino framework and going full avr-libc.  
The benefits in my opinion are smaller size, cleaner code, less dependencies, less memory usage, better performance, good learning opportunity for some of us.

I'm all ears!

SchemingWeasels committed:
Improved name scrolling

SchemingWeasels commented on Embedded/MaxMix/Display.ino:
```diff
@@ -88,23 +89,30 @@ void DisplayGameSelectScreen(Adafruit_SSD1306* display, char* name, uint8_t volu
 
+  // Mask texts
+  display->fillRect(0, 0, DISPLAY_AREA_CENTER_MARGIN_SIDE, DISPLAY_CHAR_HEIGHT_X1, 0);
+  display->fillRect(DISPLAY_WIDTH - DISPLAY_AREA_CENTER_MARGIN_SIDE + 1, 0, DISPLAY_WIDTH, DISPLAY_CHAR_HEIGHT_X1, 0);
```
The +1 here is somewhat a "magic number" but I'm not sure if it makes sense to define a const just for this. 
The +1 here is caused by the side margin (DISPLAY_AREA_CENTER_MARGINE_SIDE = 11) max char width (DISPLAY_CHAR_MAX_WIDTH_X1 = 107) not adding up to the display width (128). 

The way I thought it should work would be DISPLAY_WIDTH = 2 (MARGIN_SIDE) + CHAR_MAX_WIDTH
But currently, if you calculate this you'll get 128 = 129, and thus the +1 to offset the mask.

Thoughts?

SchemingWeasels reviewed (commented):
Fixed the bug and redid the way scrolling works. I think it's a whole lot better now, what do you think?
Using 'now' from loop instead of calling millis() when handling scrolling.
Changed the default scroll speed to 20/30.
Didn't change the default sleep time. I think it should be a separate PR if we really want to change it.

XScorpion2 commented on Embedded/MaxMix/Display.ino:
```diff
@@ -88,23 +89,30 @@ void DisplayGameSelectScreen(Adafruit_SSD1306* display, char* name, uint8_t volu
-void DisplayApplicationEditScreen(Adafruit_SSD1306* display, char* name, uint8_t volume, bool isMuted, uint8_t modeIndex, uint8_t modeCount)
+void DisplayApplicationEditScreen(Adafruit_SSD1306* display, char* name, uint8_t volume, bool isMuted, uint8_t modeIndex, uint8_t modeCount, uint32_t* displayScrollTimer, uint32_t now)
 {
   display->clearDisplay();
```
calculate the new scrollOffset before clearing the display. If the scroll offset has not changed from last call, return. If it has, store the results and continue.

XScorpion2 replied:
Note: This should be done in the other locations as well where you are scrolling.

XScorpion2 commented:
So the main performance hit by this change is going to be flushing the display every loop. Adding a check based on the calculated scroll offset before doing any display code, and exiting if it is the same as last call will restore quite a bit of performance. From there I would say if the performance is good (IE: using it feels fine), then to merge it and I'll continue with my performance changes in Display.ino, and possible make modifications to the Adafruit library as well. Based on the usage of the library, I think my oled driver code will be a poor fit as it was designed with different restrictions and requirements in mind.

t3knomanzer commented:
Alright then. I’ll test it in a few hours and merge if it’s fine. Can you guys test it on your side too?

SchemingWeasels committed:
Optimized display drawing

Screen is only drawn when absolutely needed.
isDirty triggers a redraw regardless of other factors.

SchemingWeasels commented:
The display only redraws when it has to now, this change drastically reduces the number of display updates.  
Didn't notice a performance issue before, so I couldn't feel any difference with this change either. It should work exactly like before unless I missed something.

t3knomanzer commented:
@SchemingWeasels I'm not sure about the reverse scroll. I don't find it useful reading backwards really. Personally I prefer the original implementation where it just reset back after done scrolling.

Thoughts?

Also, there is a bug in Game mode, the acceleration doesn't work. It seems to be caused by the reverse direction detection code, since both items move in opposite directions.

I can take care of that, no worries.

```
else if (stateGame == STATE_GAME_EDIT)
    {
      items[itemIndexGameA].volume = ComputeAcceleratedVolume(encoderDelta, deltaTime, items[itemIndexGameA].volume);
      items[itemIndexGameB].volume = ComputeAcceleratedVolume(-encoderDelta, deltaTime, items[itemIndexGameB].volume);

      SendItemVolumeCommand(&items[itemIndexGameA], sendBuffer, encodeBuffer);
      SendItemVolumeCommand(&items[itemIndexGameB], sendBuffer, encodeBuffer);
    }
```

t3knomanzer commented:
@SchemingWeasels I have the acceleration fixed in another PR. As soon as you finish this one I'll merge it, rebase the fix and merge it too, then I'll trigger the build for the 1.1.0 release.

SchemingWeasels commented:
@t3knomanzer I found that the way the item name just suddenly appears to be very jarring. I have another idea. I'm thinking of scrolling past the last char and looping back around to the start of the name. Sorta "continuous scroll" for the name.

For example 
[SPOTIFY PREM]
[POTIFY PREMI]
[OTIFY PREMIU]
:
[PREMIUM    SP]
[REMIUM   SPO]
[EMIUM   SPOT]
:
[SPOTIFY PREM]

Then we'll be back to 2 delays only, one for intervals between scrolls and the other is the delay between steps. Not sure how I wanna implement this yet though. Perhaps some coffee would bring ideas. 

Do you think this would be more intuitive?

t3knomanzer commented:
Absolutely, that’s be the perfect solution. Scroll all the way to the initial position the pause for a bit and scroll again. 

Thanks for the work!

SchemingWeasels committed:
Updated scrolling behaviour

SchemingWeasels commented on Embedded/MaxMix/Display.ino:
```diff
@@ -333,5 +430,51 @@ void DrawSelectionVolumeBar(Adafruit_SSD1306* display, uint8_t volume, bool isMu
+  strcat_P(displayName, DISPLAY_SCROLL_ITEM_SEPARATOR);
+  
+  uint16_t maxOffset;
+  display->getTextBounds(displayName, 0, 0, NULL, NULL, &maxOffset, NULL);
```
So the spacing between characters is not constant, it depends on what the characters are, the font size, and font style etc. Calculating manually will cause some weird edge cases to return wrong offset/calculation values. For example, applications with multiple symbols or special characters like tab and space.

To keep things compatible with any app name even weird ones, which I think is a fair assumption that there can and will be weird names, I used **display->getTextBounds** which is part of the GFX lib.

SchemingWeasels reviewed (commented):
Started running into SRAM issues. I put all the consts into flash and that helped a bit. I ran into SRAM issues when I joined the names back to back into a long string.

In my opinion, shouldn't most of these be precompiler directives (#define) ? 

Also, I made the separator customizable. It's currently 3 spaces, but I tried with some other chars and it looked alright as well.

Honestly, I'm not really stoked on the way I did it. Feels really sloppy, evident by  the running out of SRAM.

t3knomanzer commented:
Why don’t you just draw the same string twice instead of modifying the string itself?

Ruben Henares

> On Aug 16, 2020, at 08:30, SchemingWeasels <notifications@github.com> wrote:
> 
> ﻿
> @SchemingWeasels commented on this pull request.
> 
> Started running into SRAM issues. I put all the consts into flash and that helped a bit. I ran into SRAM issues when I joined the names back to back into a long string.
> 
> In my opinion, shouldn't most of these be precompiler directives (#define) ?
> 
> Also, I made the separator customizable. It's currently 3 spaces, but I tried with some other chars and it looked alright as well.
> 
> Honestly, I'm not really stoked on the way I did it. Feels really sloppy, evident by the running out of SRAM.
> 
> In Embedded/MaxMix/Display.ino:
> 
> > +      UpdateScrollTimer();
> +    }
> +    return 0;
> +  }
> +}
> +
> +//---------------------------------------------------------
> +//---------------------------------------------------------
> +uint16_t GetMaxScrollOffset(Adafruit_SSD1306* display, char* name)
> +{
> +  char displayName[strlen(name) + strlen(DISPLAY_SCROLL_ITEM_SEPARATOR) + strlen(name) + 1] = "";
> +  strcpy(displayName, name);
> +  strcat_P(displayName, DISPLAY_SCROLL_ITEM_SEPARATOR);
> +  
> +  uint16_t maxOffset;
> +  display->getTextBounds(displayName, 0, 0, NULL, NULL, &maxOffset, NULL);
> So the spacing between characters is not constant, it depends on what the characters are, the font size, and font style etc. Calculating manually will cause some weird edge cases to return wrong offset/calculation values. For example, applications with multiple symbols or special characters like and .
> 
> To keep things compatible with any app name even weird ones, which I think is a fair assumption that there can and will be weird names, I used display->getTextBounds which is part of the GFX lib.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

SchemingWeasels commented:
@t3knomanzer I do. I modify the string to parse into getTextBounds. Drawing is just using the same string twice if the name is scrollable.

t3knomanzer commented:
Sorry I haven’t been able to look at this. I’ve had some unexpected things to take care of today.

t3knomanzer commented:
Hey,

First of all, sorry for the delay, I've had some unexpected things to deal with.
After reviewing this I think that we've iterated on this so much that's become a little messy so I took the ideas from this implementation and re-implemented it from scratch in PR #141.

Thank you all for the awesome work!
